### DISCUSSION TITLE
Composition Syntactic Sugar

### DISCUSSION THREAD

**Discussion by 11clock** [dotnet/csharplang#6375] [Language Ideas]
I often use the following pattern when doing a compositional approach in my code.

```cs
public class A
{
    public void Method(int arg1, int arg2, int arg3) {}
}

public class B
{
    private A _a = new();

    public void Method(int arg1, int arg2, int arg3) => _a.Method(arg1, arg2, arg3);
}
```

It gets a bit tedious to make methods that only just refer to another instance's method. I would like a more concise way to do it. The following is one idea using a new "mixin" keyword.

```cs
public class A
{
    public void Method(int arg1, int arg2, int arg3) {}
}

public class B
{
    private A _a = new();

    public mixin Method => _a;
}
```

One concern is if _a in this case is null. We could either let it throw an error, or probably have the ability to support null safety.

```cs
public mixin Method => _a?;
```

I am open to other suggestions as well.

**Comment by HaloFour** [comment]
This is something you could accomplish with [source generators](https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview) where the generator copies the members of the target type and delegates to an instance.

```cs
public partial class B
{
    [Mixin] private readonly A _a;
}

// generated code
public partial class B
{
    public void Method(int arg1, int arg2, int arg3) => _a.Method(arg1, arg2, arg3);
}
```

**Reply by 11clock** [reply]
@HaloFour What about cases where you don't want to copy all members? Say in my example class A has a Method2, but in class B I just want to expose class A's Method.

**Reply by HaloFour** [reply]
That would be up to the source generator to provide the knobs to tune exactly which members are exposed and how.  If anything the degree of policy that would apply to such a feature is likely to make it very difficult to implement as a language feature.

**Reply by En3Tho** [reply]
I've played with this exact idea, but I've called it DecoratorGenerator, not Mixin. But the idea was the same: use field as a base for decoration and then explicitly implement members you want. Everything is done by source gen
